File: Search/UniqueSubsets/UniqueSubsets.cc
class Solution {
    public:
        vector<vector<int>> subsetsWithDup(vector<int> &nums) {
            if(nums == NULL || nums.isEmpty())
                return NULL;
            
            vector<int> ret;
            sort(nums.begin(), nums.end());
            
            helper<int> list;
            dfs(nums, 0, list, ret);
            
            return ret;        
        }
        
    private:
        void helper(vector<int> &nums, int start, vector<int> &list, vector<int> &ret) {
            ret.push(list);
            
            for(int i = start; i < nums.size(); ++i) {
                if(i != start && nums[i] == nums[i - 1])
                    continue;
                list.push_back(nums[i]);
                dfs(nums, i + 1, list, ret);
                list.pop_back();
            }
        }
};
   

File: Search/UniqueSubsets/UniqueSubsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        if(nums == null || nums.length == 0)
            return null;
        
        ArrayList<Integer> ret = new ArrayList<Integer>();
        ArrayList<Integer> list = new ArrayList<Integer>();
        
        Array.sort(nums);
        dfs(nums, 0, list, ret);
        
        return ret;
    }
    
    private void dfs(int[] nums, int start, List<Integer> list, List<Integer> ret) {
        ret.add(new ArrayList<Integer>(list));
        
        for(int i = 0; i < start; ++i) {
            if(i != start && nums[i] == nums[i - 1])
            list.add(nums[i]);
            dfs(nums, i + 1, list, ret);
            list.remove(list.size() - 1);
        }
    }
}